Add exclude-config option

galderz opened the pull request:
This PR Replaces #3149.

It uses a more fine grained approach to exclude configuration from jars. A new option `--exclude-config` is added that takes as a value a comma separating a jar name and regular expression of resources to exclude from the given jar. Jar names can include path elements for situations where two jar names have the same name but different leading path.


Examples:
```
--exclude-config foo.jar,META-INF\/native-image\/.*.properties
--exclude-config bar.jar,META-INF\/native-image\/.*.json
--exclude-config foo/example.jar,META-INF\/native-image\/.*.json
--exclude-config bar/example.jar,META-INF\/native-image\/.*.json
```

The approach was agreed with @olpaw on [this slack chat](https://graalvm.slack.com/archives/CN9KSFB40/p1611664256033900).

maxandersen commented:
Is there way to use pattern for the jar?

galderz commented:
> Is there way to use pattern for the jar?

Not at the moment, but could be added.

olpaw commented:
> > Is there way to use pattern for the jar?
> 
> Not at the moment, but could be added.

Be care how you add that so that the separating of the regex for the jarfile and the regex for the resource path is unambiguous. I.e. if the regex of the path would contain a `,` you might mistakenly interpret it as the separator of jarfile-regex and resourcepath-regex. (I know ... `,` is quite unlikely in a regex for a path, but you never know what users are up to). Maybe the easiest solution would be to make the option take two arguments ... like
```
--exclude-config foo/.*\.jar META-INF\/native-image\/.*.json
```

olpaw commented on substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java:
```diff
@@ -151,6 +152,18 @@ public boolean consume(ArgumentQueue args) {
+                if (excludeConfig == null) {
+                    NativeImage.showError(headArg + " requires a jar and resource regular expression, comma separated");
+                }
+                final String[] split = excludeConfig.split(",");
```
If your regex contains a `","` you will only use the first half of it as the second argument to nativeImage.addExcludeConfig. What you need here is
```java
String[] split = excludeConfig.split(",", 2);
```

olpaw commented:
> > Is there way to use pattern for the jar?
> 
> Not at the moment, but could be added.

@galderz are you going to change the PR to make the path matching also be based on a regex (as @maxandersen suggested) or do you consider this PR ready as is?

olpaw reviewed (changes requested):
See https://github.com/oracle/graal/pull/3179/files#r571867459

galderz commented:
I'm handling this today. I'm going to try to apply @maxandersen's suggestion, along with the split into 2 parameters that @olpaw suggested. That sounds like a good combination.

galderz commented:
Updated PR. Added a commit to switch to 2 parameters and hence enable jar files to also be regular expressions. I can squash commits when it's ready to go in.

olpaw reviewed (approved):
LGTM. Please squash the commit so that I can create an internal PR.

galderz commented:
Thx @olpaw. Squashed and pushed.

zakkak commented:
@galderz it looks like there is a [formatting issue](https://github.com/oracle/graal/runs/1917831543?check_suite_focus=true#step:8:1603)

olpaw commented:
> @galderz it looks like there is a [formatting issue](https://github.com/oracle/graal/runs/1917831543?check_suite_focus=true#step:8:1603)

@zakkak is right. This should be fixed before I try to push this to master.
@galderz please adjust the formatting accordingly.

galderz commented:
Looking into the formatting issue

galderz commented:
How reliable is this formatting automation? When I run `mx eclipseformat -e /opt/eclipse/Eclipse.app/Contents/MacOS/eclipse --primary` locally I see [all these files formatted](https://gist.github.com/galderz/7c00ad5ca67ad6b88afc6d08f335bd15).

galderz commented:
Pushed an updated. Formatting should be fixed.

olpaw commented:
Your PR passed our gates and is now in the merge queue. ETA Monday
